File: Two Pointers/438.find-all-anagrams-in-a-string.py
#
# @lc app=leetcode id=438 lang=python3
#
# [438] Find All Anagrams in a String
#

from typing import List
import collections

# @lc code=start
class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        d = collections.Counter(p)
        j = 0
        res = []
        for i, ch in enumerate(s):
            # Shorten the sliding Window when the size crosses
            if i-j >= len(p):
                if s[j] in d:
                    d[s[j]] += 1
                j += 1
            # Adding the chracter to dictionary
            if ch in d:
                d[ch] -= 1
            flag = True
            for val in d.keys():
                if d[val] != 0:
                    flag = False
                    break
            if flag:
                res.append(j)
        return res
# @lc code=end

File: Sliding Window/438.find-all-anagrams-in-a-string.cpp
/*
 * @lc app=leetcode id=438 lang=cpp
 *
 * [438] Find All Anagrams in a String
 */

#include <bits/stdc++.h>
using namespace std;

// @lc code=start
class Solution {
   public:
    vector<int> findAnagrams(string s, string p) {
        // Very similar idea as problem 567. Permutation in string
        vector<int> ans;
        unordered_map<char, int> mp;
        for (char &ch : p) mp[ch]++;
        int unique = mp.size();
        int l = 0;
        for (int r = 0; r < s.size(); r++) {
            if (mp.count(s[r])) {
                mp[s[r]]--;
                if (mp[s[r]] == 0) unique--;
            }
            if (r - l + 1 == p.size()) {
                if (unique == 0) ans.push_back(l);
                if(mp.count(s[l])){
                    mp[s[l]]++;
                    if(mp[s[l]] == 1) unique++;
                }
                l++;
            }
        }
        return ans;
    }
};
// @lc code=end
